In [2]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

# import numpy as np # linear algebra
# import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# # Input data files are available in the read-only "../input/" directory
# # For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

# import os
# for dirname, _, filenames in os.walk('/kaggle/input'):
#     for filename in filenames:
#         print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/spaceship-titanic/sample_submission.csv
/kaggle/input/spaceship-titanic/train.csv
/kaggle/input/spaceship-titanic/test.csv


In [1]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity="all"

In [2]:
import numpy as np
import pandas as pd

df = pd.read_csv('C:/ds_work/source_code/Project_DS/data/train.csv')

In [3]:
df.groupby('Destination').count()
df.groupby('HomePlanet').count()
print('Cabin의 개수',len(df.groupby('Cabin').count()))

PassengerId  HomePlanet  CryoSleep  Cabin   Age   VIP  \
Destination                                                            
55 Cancri e           1800        1769       1756   1750  1766  1757   
PSO J318.5-22          796         780        777    778   782   774   
TRAPPIST-1e           5915        5765       5765   5788  5787  5781   

               RoomService  FoodCourt  ShoppingMall   Spa  VRDeck  Name  \
Destination                                                               
55 Cancri e           1771       1763          1752  1761    1760  1757   
PSO J318.5-22          775        780           778   780     784   779   
TRAPPIST-1e           5787       5791          5778  5790    5781  5779   

               Transported  
Destination                 
55 Cancri e           1800  
PSO J318.5-22          796  
TRAPPIST-1e           5915

PassengerId  CryoSleep  Cabin  Destination   Age   VIP  \
HomePlanet                                                           
Earth              4602       4488   4507         4503  4514  4487   
Europa             2131       2073   2070         2094  2085  2089   
Mars               1759       1716   1722         1717  1716  1716   

            RoomService  FoodCourt  ShoppingMall   Spa  VRDeck  Name  \
HomePlanet                                                             
Earth              4499       4505          4490  4503    4491  4496   
Europa             2095       2085          2078  2086    2094  2085   
Mars               1723       1721          1723  1723    1720  1718   

            Transported  
HomePlanet               
Earth              4602  
Europa             2131  
Mars               1759

Cabin의 개수 6560


컬럼
- Destination 목적지 {55 Cancri e, PSO J318.5-22, TRAPPIST-1e}
- HomePlanet  {Earth , Europa, Mars}
- Cabin의 종류 6560
    - Cabin을 앞 대문자만 남기면 의미가 있을 수 있을까?
    - 돈이 많은 사람이 사는 자리 등으로 의미가 있는지확인
    - A,B,C,D,E,F,G,T로 변환 후 다시 숫자로 0~7까지 변환
- True / False 
    - CryoSleep
    - VIP
    - Transported(target)
- 삭제
    - PassengerId
    - Name

In [4]:
# 범주형 데이터 변환
def df_replace(dataframe):

    global df_train
    df = dataframe
    df = df.replace({'55 Cancri e':0,'PSO J318.5-22':1,'TRAPPIST-1e':2})
    df = df.replace({'Earth':0,'Europa':1,'Mars':2})
    df = df.replace({False:0,True:1})

    df['Cabin'] = df['Cabin'].str[0]
    df = df.replace({'A':0,'B':1,'C':2,'D':3,'E':4,'F':5,'G':6,'T':7})


    ## 범주형 데이터 확인할 수 있게 dict로 만들어 놓은 것
    Cabin = {'A':0,'B':1,'C':2,'D':3,'E':4,'F':5,'G':6,'T':7}
    Destination = {'55 Cancri e': 0, 'PSO J318.5-22': 1,'TRAPPIST-1e':2}
    HomePlanet = {'Earth':0,'Europa':1,'Mars':2}
    CryoSleep = {'False':0,'True':1}
    VIP = {'False':0,'True':1}
    Transported = {'False':0,'True':1}


    df_train = df.drop(['PassengerId','Name'],axis=1)
    # df 전체 float형으로 바꾸기
    df_train = df_train.astype(float)

df_replace(df)

In [5]:
df_train

HomePlanet  CryoSleep  Cabin  Destination   Age  VIP  RoomService  \
0            1.0        0.0    1.0          2.0  39.0  0.0          0.0   
1            0.0        0.0    5.0          2.0  24.0  0.0        109.0   
2            1.0        0.0    0.0          2.0  58.0  1.0         43.0   
3            1.0        0.0    0.0          2.0  33.0  0.0          0.0   
4            0.0        0.0    5.0          2.0  16.0  0.0        303.0   
...          ...        ...    ...          ...   ...  ...          ...   
8688         1.0        0.0    0.0          0.0  41.0  1.0          0.0   
8689         0.0        1.0    6.0          1.0  18.0  0.0          0.0   
8690         0.0        0.0    6.0          2.0  26.0  0.0          0.0   
8691         1.0        0.0    4.0          0.0  32.0  0.0          0.0   
8692         1.0        0.0    4.0          2.0  44.0  0.0        126.0   

      FoodCourt  ShoppingMall     Spa  VRDeck  Transported  
0           0.0           0.0     0.0     0.0          0.0  
1           9.0          25.0   549.0    44.0          1.0  
2        3576.0           0.0  6715.0    49.0          0.0  
3        1283.0         371.0  3329.0   193.0          0.0  
4          70.0         151.0   565.0     2.0          1.0  
...         ...           ...     ...     ...          ...  
8688     6819.0           0.0  1643.0    74.0          0.0  
8689        0.0           0.0     0.0     0.0          0.0  
8690        0.0        1872.0     1.0     0.0          1.0  
8691     1049.0           0.0   353.0  3235.0          0.0  
8692     4688.0           0.0     0.0    12.0          1.0  

[8693 rows x 12 columns]

### null값 처리
XGBOOST를 사용해서 pridct값을 null값으로 대체

In [6]:
# df_homeplanet =  df_train중에서 HomePlanet의 null값이 존재하는 열만 추출 
# -> 훈련된 모델의 값으로 다시 채울 예정
df_FoodCourt = df_train[df_train['FoodCourt'].isnull()]

#df_train 중에서 HomePlanet의 null값이 존재하는 열만 추출 -> 추후 훈련할 데이터프레임
df_train01 = df_train.dropna(subset=['FoodCourt'])
# df_train
df_train[df_train['FoodCourt'].isnull()]

HomePlanet  CryoSleep  Cabin  Destination   Age  VIP  RoomService  \
95           0.0        0.0    5.0          2.0  24.0  0.0          0.0   
218          0.0        1.0    6.0          1.0  27.0  0.0          0.0   
367          0.0        0.0    5.0          0.0  17.0  0.0        177.0   
385          1.0        0.0    4.0          2.0  31.0  0.0        417.0   
456          1.0        1.0    NaN          0.0  27.0  0.0          0.0   
...          ...        ...    ...          ...   ...  ...          ...   
8396         0.0        0.0    5.0          2.0  27.0  0.0          1.0   
8422         0.0        1.0    5.0          0.0  15.0  0.0          0.0   
8460         1.0        0.0    3.0          2.0  60.0  0.0          0.0   
8537         0.0        0.0    5.0          2.0  26.0  0.0          0.0   
8623         1.0        0.0    2.0          0.0  30.0  0.0       1569.0   

      FoodCourt  ShoppingMall    Spa  VRDeck  Transported  
95          NaN           0.0    0.0    17.0          1.0  
218         NaN           0.0    0.0     0.0          0.0  
367         NaN           0.0  310.0     7.0          0.0  
385         NaN        7185.0  562.0   792.0          1.0  
456         NaN           0.0    0.0     0.0          1.0  
...         ...           ...    ...     ...          ...  
8396        NaN           0.0   20.0   948.0          0.0  
8422        NaN           0.0    0.0     0.0          1.0  
8460        NaN         947.0  668.0     1.0          1.0  
8537        NaN           0.0    1.0    48.0          1.0  
8623        NaN           0.0    NaN  2252.0          0.0  

[183 rows x 12 columns]

In [7]:
# train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(df_train01.drop('FoodCourt',axis=1),df_train01['FoodCourt'], test_size=0.2, random_state=42)

# scaler = StandardScaler().fit(X_train)
# X_train = scaler.transform(X_train)
# X_test = scaler.transform(X_test)
X_train.shape
y_train.shape
X_test.shape
y_test.shape

(6808, 11)

(6808,)

(1702, 11)

(1702,)

In [13]:
X_test.isnull().sum()

HomePlanet      35
CryoSleep       50
Cabin           44
Destination     41
Age             34
VIP             42
RoomService     43
ShoppingMall    41
Spa             28
VRDeck          44
Transported      0
dtype: int64

In [16]:
y_test.isnull().sum()

0

In [24]:
X_train.isnull().sum()

HomePlanet      164
CryoSleep       164
Cabin           150
Destination     135
Age             141
VIP             157
RoomService     136
ShoppingMall    165
Spa             150
VRDeck          138
Transported       0
dtype: int64

In [19]:
#XGBoost1
import xgboost as xgb
import pandas as pd
import seaborn as sns
from xgboost import XGBRegressor
from sklearn.metrics import accuracy_score

model = xgb.XGBRegressor(early_stopping_rounds=10)
#eval_set = [(X_test, y_test)]
model.fit(X_train, y_train, eval_set=eval_set, verbose=False)#, verbose=True

pred = model.predict(X_test)
acc = accuracy_score(y_test,pred)
acc

# from sklearn.metrics import confusion_matrix
# import scikitplot as skplt

# pred = model.predict(X_test)
# # 실제 , 예측
# cm = confusion_matrix(y_test, pred)
# print('confusion_matrix : \n', cm)

# skplt.metrics.plot_confusion_matrix(y_test,pred,figsize=(8,6))
# plt.show()

# cl_report = metrics.classification_report(y_test,pred)
# print('리포트:\n',cl_report)

XGBRegressor(base_score=0.5, booster='gbtree', callbacks=None,
             colsample_bylevel=1, colsample_bynode=1, colsample_bytree=1,
             early_stopping_rounds=10, enable_categorical=False,
             eval_metric=None, feature_types=None, gamma=0, gpu_id=-1,
             grow_policy='depthwise', importance_type=None,
             interaction_constraints='', learning_rate=0.300000012, max_bin=256,
             max_cat_threshold=64, max_cat_to_onehot=4, max_delta_step=0,
             max_depth=6, max_leaves=0, min_child_weight=1, missing=nan,
             monotone_constraints='()', n_estimators=100, n_jobs=0,
             num_parallel_tree=1, predictor='auto', random_state=0, ...)

ValueError: Classification metrics can't handle a mix of multiclass and continuous targets

In [23]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_train, y_train)

ValueError: Input contains NaN, infinity or a value too large for dtype('float64').

In [21]:
## df_01 은 df_train 중에서 HomePlanet의 값이 null인 경우의 데이터프레임
df_01 = df_FoodCourt.drop('FoodCourt',axis=1)
df_01

pred = model.predict(df_01)
pred


df_FoodCourt['FoodCourt'] = pred
df_FoodCourt
df_update = pd.concat([df_train01,df_homeplanet],axis=0)
df_update

HomePlanet  CryoSleep  Cabin  Destination   Age  VIP  RoomService  \
95           0.0        0.0    5.0          2.0  24.0  0.0          0.0   
218          0.0        1.0    6.0          1.0  27.0  0.0          0.0   
367          0.0        0.0    5.0          0.0  17.0  0.0        177.0   
385          1.0        0.0    4.0          2.0  31.0  0.0        417.0   
456          1.0        1.0    NaN          0.0  27.0  0.0          0.0   
...          ...        ...    ...          ...   ...  ...          ...   
8396         0.0        0.0    5.0          2.0  27.0  0.0          1.0   
8422         0.0        1.0    5.0          0.0  15.0  0.0          0.0   
8460         1.0        0.0    3.0          2.0  60.0  0.0          0.0   
8537         0.0        0.0    5.0          2.0  26.0  0.0          0.0   
8623         1.0        0.0    2.0          0.0  30.0  0.0       1569.0   

      ShoppingMall    Spa  VRDeck  Transported  
95             0.0    0.0    17.0          1.0  
218            0.0    0.0     0.0          0.0  
367            0.0  310.0     7.0          0.0  
385         7185.0  562.0   792.0          1.0  
456            0.0    0.0     0.0          1.0  
...            ...    ...     ...          ...  
8396           0.0   20.0   948.0          0.0  
8422           0.0    0.0     0.0          1.0  
8460         947.0  668.0     1.0          1.0  
8537           0.0    1.0    48.0          1.0  
8623           0.0    NaN  2252.0          0.0  

[183 rows x 11 columns]

ValueError: Input contains NaN, infinity or a value too large for dtype('float64').

In [22]:
df_update[df_update['FoodCourt'].isnull()]

NameError: name 'df_update' is not defined